sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/objects.scala

```diff
@@ -96,3 +98,22 @@ object CombineTypedFilters extends Rule[LogicalPlan] {
     }
   }
 }
+
+/**
+ * Removes MapObjects when the following conditions are satisfied
+ *   1. Mapobject(e) where e is lambdavariable(), which means types for input output
+ *      are primitive types
+ *   2. no custom collection class specified
+ *      representation of data item.  For example back to back map operations.
+ */
+object EliminateMapObjects extends Rule[LogicalPlan] {
+  def apply(plan: LogicalPlan): LogicalPlan = plan transform {
+    case DeserializeToObject(Invoke(
+        MapObjects(_, _, _, _ : LambdaVariable, inputData, None),
```

Comment:
When we run the following code, the following exception occurs. This is because `UnsafeArrayData.copy()`, which is unsupported, is called.

```
    val ds = Seq((1, Map(2 -> 3))).toDS.map(t => t)
    ds.collect.toSeq
```